I didn't expect the `ZLIB_COMPAT`run to fail with `/usr/bin/ld: cannot find -lzlibstatic: No such file or directory`... This kind of revealed another bug...

--- a/.github/workflows/cmake.yml
+++ b/.github/workflows/cmake.yml
@@ -295,6 +295,24 @@ jobs:
             gcov-exec: llvm-cov-11 gcov
             codecov: ubuntu_clang
 
+          # Check for undefined symbols in the version script for the modern api
+          - name: Ubuntu Clang Undefined Symbols
+            os: ubuntu-latest
+            compiler: clang-11
+            cxx-compiler: clang++-11
+            cmake-args: -DCMAKE_SHARED_LINKER_FLAGS="-fuse-ld=lld -Wl,--no-undefined-version" -DZLIBNG_ENABLE_TESTS=OFF
+            build-shared: ON
+            packages: clang-11 llvm-11 lld
+
+          # Check for undefined symbols in the version script for the compat api
+          - name: Ubuntu Clang Undefined Symbols Compat
+            os: ubuntu-latest
+            compiler: clang-11
+            cxx-compiler: clang++-11
+            cmake-args: -DCMAKE_SHARED_LINKER_FLAGS="-fuse-ld=lld -Wl,--no-undefined-version" -DZLIBNG_ENABLE_TESTS=OFF -DZLIB_COMPAT=ON
+            build-shared: ON
+            packages: clang-11 llvm-11 lld
+
           - name: Ubuntu Clang Inflate Strict
             os: ubuntu-latest
             compiler: clang-11
@@ -518,7 +536,7 @@ jobs:
       run: |
         cmake -S ${{ matrix.build-src-dir || '.' }} -B ${{ matrix.build-dir || '.' }} ${{ matrix.cmake-args }} \
           -DCMAKE_BUILD_TYPE=${{ matrix.build-config || 'Release' }} \
-          -DBUILD_SHARED_LIBS=OFF \
+          -DBUILD_SHARED_LIBS=${{ matrix.build-shared || 'OFF' }} \
           -DWITH_FUZZERS=ON \
           -DWITH_MAINTAINER_WARNINGS=ON \
           ${{ matrix.codecov && '-DWITH_CODE_COVERAGE=ON' }}

--- a/test/CMakeLists.txt
+++ b/test/CMakeLists.txt
@@ -230,12 +230,15 @@ if(WITH_GTEST)
         target_link_libraries(gtest_zlib external_zlib)
     endif()
 
-    if((NOT ZLIBNG_ENABLE_TESTS) AND ZLIB_COMPAT AND (NOT DEFINED BUILD_SHARED_LIBS OR BUILD_SHARED_LIBS))
-        # Link dynamically in order to be able to substitute zlib-ng with zlib.
-        target_link_libraries(gtest_zlib zlib GTest::Main)
-    else()
+    if(NOT DEFINED BUILD_SHARED_LIBS)
         # Link statically in order to test internal zlib-ng functions.
         target_link_libraries(gtest_zlib zlibstatic)
+    else()
+        target_link_libraries(gtest_zlib zlib)
+    endif()
+
+    if(BUILD_SHARED_LIBS)
+        target_link_libraries(gtest_zlib GTest::Main)
     endif()
     target_link_libraries(gtest_zlib GTest::GTest)
 

--- a/test/fuzz/CMakeLists.txt
+++ b/test/fuzz/CMakeLists.txt
@@ -31,7 +31,12 @@ foreach(FUZZER ${FUZZERS})
         target_sources(${FUZZER} PRIVATE standalone_fuzz_target_runner.c)
     endif()
 
-    target_link_libraries(${FUZZER} zlibstatic)
+    if(NOT DEFINED BUILD_SHARED_LIBS)
+        target_link_libraries(${FUZZER} zlibstatic)
+    else()
+        target_link_libraries(${FUZZER} zlib)
+    endif()
+
     if(FUZZING_ENGINE_FOUND)
         target_link_libraries(${FUZZER} ${FUZZING_ENGINE})
     endif()

